File: hash_table.py
import random

# ============================================================
#   HASH TABLE - Tugas #8: Hashing in Data Structure
#   Teknik Collision: Chaining (Separate Chaining)
#   Nama: [Isi Nama Anda]
# ============================================================

TABLE_SIZE = 127  # Ukuran hash table (bilangan prima agar distribusi merata)


class HashTable:
    def __init__(self, size):
        self.size = size
        # Setiap slot berupa list (chain) untuk menangani collision
        self.table = [[] for _ in range(size)]
        self.total_data = 0

    # ----------------------------------------------------------
    # HASH FUNCTION: modulo dengan ukuran tabel (bilangan prima)
    # ----------------------------------------------------------
    def hash_function(self, key):
        return key % self.size

    # ----------------------------------------------------------
    # 1. INPUT DATA
    # ----------------------------------------------------------
    def insert(self, key):
        index = self.hash_function(key)
        chain = self.table[index]

        # Cek duplikat
        if key in chain:
            print(f"  [!] Data {key} sudah ada di tabel (duplikat ditolak).")
            return False

        chain.append(key)
        self.total_data += 1
        print(f"  [+] Data {key} berhasil dimasukkan -> index [{index}]")
        return True

    # ----------------------------------------------------------
    # 2. HAPUS DATA
    # ----------------------------------------------------------
    def delete(self, key):
        index = self.hash_function(key)
        chain = self.table[index]

        if key in chain:
            chain.remove(key)
            self.total_data -= 1
            print(f"  [-] Data {key} berhasil dihapus dari index [{index}].")
            return True
        else:
            print(f"  [!] Data {key} tidak ditemukan. Tidak ada yang dihapus.")
            return False

    # ----------------------------------------------------------
    # 3. CARI DATA
    # ----------------------------------------------------------
    def search(self, key):
        index = self.hash_function(key)
        chain = self.table[index]

        if key in chain:
            pos = chain.index(key)
            print(f"  [✓] Data {key} DITEMUKAN -> index [{index}], posisi chain ke-{pos}.")
            return True
        else:
            print(f"  [✗] Data {key} TIDAK ditemukan di tabel.")
            return False

    # ----------------------------------------------------------
    # TAMPILKAN TABEL (ringkas: hanya slot yang terisi)
    # ----------------------------------------------------------
    def display(self):
        print("\n" + "=" * 55)
        print(f"  HASH TABLE  (size={self.size}, total data={self.total_data})")
        print("=" * 55)
        occupied = 0
        collision_slots = 0
        for i, chain in enumerate(self.table):
            if chain:
                occupied += 1
                if len(chain) > 1:
                    collision_slots += 1
                print(f"  [{i:>3}] -> {chain}")
        print("-" * 55)
        print(f"  Slot terisi     : {occupied} / {self.size}")
        print(f"  Slot collision  : {collision_slots}")
        load_factor = self.total_data / self.size
        print(f"  Load factor     : {load_factor:.3f}")
        print("=" * 55 + "\n")


# ==============================================================
#   MAIN PROGRAM
# ==============================================================
def main():
    ht = HashTable(TABLE_SIZE)

    # --- Inisialisasi: masukkan 100 angka random unik (1 – 9999) ---
    print("=" * 55)
    print("  Inisialisasi: memasukkan 100 angka random unik...")
    print("=" * 55)
    random_data = random.sample(range(1, 10000), 100)
    for num in random_data:
        ht.insert(num)

    print(f"\n  100 data berhasil dimasukkan.")
    print(f"  Data awal: {sorted(random_data)}\n")

    # --- Menu Interaktif ---
    while True:
        print("\n" + "=" * 55)
        print("         MENU HASH TABLE")
        print("=" * 55)
        print("  1. Input Data")
        print("  2. Hapus Data")
        print("  3. Cari Data")
        print("  4. Tampilkan Hash Table")
        print("  5. Keluar")
        print("=" * 55)

        pilihan = input("  Pilih menu (1-5): ").strip()

        if pilihan == "1":
            print("\n  -- INPUT DATA --")
            try:
                nilai = int(input("  Masukkan angka yang ingin diinput: "))
                ht.insert(nilai)
            except ValueError:
                print("  [!] Input tidak valid. Masukkan angka bulat.")

        elif pilihan == "2":
            print("\n  -- HAPUS DATA --")
            try:
                nilai = int(input("  Masukkan angka yang ingin dihapus: "))
                ht.delete(nilai)
            except ValueError:
                print("  [!] Input tidak valid. Masukkan angka bulat.")

        elif pilihan == "3":
            print("\n  -- CARI DATA --")
            try:
                nilai = int(input("  Masukkan angka yang ingin dicari: "))
                ht.search(nilai)
            except ValueError:
                print("  [!] Input tidak valid. Masukkan angka bulat.")

        elif pilihan == "4":
            ht.display()

        elif pilihan == "5":
            print("\n  Program selesai. Sampai jumpa!\n")
            break

        else:
            print("  [!] Pilihan tidak valid. Masukkan 1-5.")


if __name__ == "__main__":
    main()

File: HashTable.java
import java.util.*;
import java.util.LinkedList;
import java.util.Scanner;


public class HashTable {

    static final int TABLE_SIZE = 127; // Bilangan prima agar distribusi merata

    // Setiap slot berupa LinkedList untuk menangani collision (chaining)
    private LinkedList<Integer>[] table;
    private int totalData;

    // ----------------------------------------------------------
    // KONSTRUKTOR
    // ----------------------------------------------------------
    @SuppressWarnings("unchecked")
    public HashTable(int size) {
        table = new LinkedList[size];
        for (int i = 0; i < size; i++) {
            table[i] = new LinkedList<>();
        }
        totalData = 0;
    }

    // ----------------------------------------------------------
    // HASH FUNCTION: modulo dengan ukuran tabel
    // ----------------------------------------------------------
    private int hashFunction(int key) {
        return key % TABLE_SIZE;
    }

    // ----------------------------------------------------------
    // 1. INPUT DATA
    // ----------------------------------------------------------
    public boolean insert(int key) {
        int index = hashFunction(key);
        LinkedList<Integer> chain = table[index];

        // Cek duplikat
        if (chain.contains(key)) {
            System.out.println("  [!] Data " + key + " sudah ada di tabel (duplikat ditolak).");
            return false;
        }

        chain.add(key);
        totalData++;
        System.out.println("  [+] Data " + key + " berhasil dimasukkan -> index [" + index + "]");
        return true;
    }

    // ----------------------------------------------------------
    // 2. HAPUS DATA
    // ----------------------------------------------------------
    public boolean delete(int key) {
        int index = hashFunction(key);
        LinkedList<Integer> chain = table[index];

        if (chain.remove(Integer.valueOf(key))) {
            totalData--;
            System.out.println("  [-] Data " + key + " berhasil dihapus dari index [" + index + "].");
            return true;
        } else {
            System.out.println("  [!] Data " + key + " tidak ditemukan. Tidak ada yang dihapus.");
            return false;
        }
    }

    // ----------------------------------------------------------
    // 3. CARI DATA
    // ----------------------------------------------------------
    public boolean search(int key) {
        int index = hashFunction(key);
        LinkedList<Integer> chain = table[index];

        int pos = chain.indexOf(key);
        if (pos != -1) {
            System.out
                    .println("  [v] Data " + key + " DITEMUKAN -> index [" + index + "], posisi chain ke-" + pos + ".");
            return true;
        } else {
            System.out.println("  [x] Data " + key + " TIDAK ditemukan di tabel.");
            return false;
        }
    }

    // ----------------------------------------------------------
    // TAMPILKAN TABEL (hanya slot yang terisi)
    // ----------------------------------------------------------
    public void display() {
        System.out.println("\n" + "=".repeat(55));
        System.out.printf("  HASH TABLE  (size=%d, total data=%d)%n", TABLE_SIZE, totalData);
        System.out.println("=".repeat(55));

        int occupied = 0;
        int collisionSlots = 0;

        for (int i = 0; i < TABLE_SIZE; i++) {
            if (!table[i].isEmpty()) {
                occupied++;
                if (table[i].size() > 1)
                    collisionSlots++;
                System.out.printf("  [%3d] -> %s%n", i, table[i]);
            }
        }

        System.out.println("-".repeat(55));
        System.out.printf("  Slot terisi    : %d / %d%n", occupied, TABLE_SIZE);
        System.out.printf("  Slot collision : %d%n", collisionSlots);
        System.out.printf("  Load factor    : %.3f%n", (double) totalData / TABLE_SIZE);
        System.out.println("=".repeat(55) + "\n");
    }

    // ==============================================================
    // MAIN PROGRAM
    // ==============================================================
    public static void main(String[] args) {
        HashTable ht = new HashTable(TABLE_SIZE);
        Scanner scanner = new Scanner(System.in);

        // --- Inisialisasi: masukkan 100 angka random unik (1 – 9999) ---
        System.out.println("=".repeat(55));
        System.out.println("  Inisialisasi: memasukkan 100 angka random unik...");
        System.out.println("=".repeat(55));

        List<Integer> pool = new ArrayList<>();
        for (int i = 1; i <= 9999; i++)
            pool.add(i);
        Collections.shuffle(pool);
        List<Integer> randomData = pool.subList(0, 100);

        for (int num : randomData) {
            ht.insert(num);
        }

        List<Integer> sorted = new ArrayList<>(randomData);
        Collections.sort(sorted);
        System.out.println("\n  100 data berhasil dimasukkan.");
        System.out.println("  Data awal: " + sorted + "\n");

        // --- Menu Interaktif ---
        while (true) {
            System.out.println("\n" + "=".repeat(55));
            System.out.println("         MENU HASH TABLE");
            System.out.println("=".repeat(55));
            System.out.println("  1. Input Data");
            System.out.println("  2. Hapus Data");
            System.out.println("  3. Cari Data");
            System.out.println("  4. Tampilkan Hash Table");
            System.out.println("  5. Keluar");
            System.out.println("=".repeat(55));
            System.out.print("  Pilih menu (1-5): ");

            String pilihan = scanner.nextLine().trim();

            switch (pilihan) {
                case "1":
                    System.out.println("\n  -- INPUT DATA --");
                    System.out.print("  Masukkan angka yang ingin diinput: ");
                    try {
                        int nilai = Integer.parseInt(scanner.nextLine().trim());
                        ht.insert(nilai);
                    } catch (NumberFormatException e) {
                        System.out.println("  [!] Input tidak valid. Masukkan angka bulat.");
                    }
                    break;

                case "2":
                    System.out.println("\n  -- HAPUS DATA --");
                    System.out.print("  Masukkan angka yang ingin dihapus: ");
                    try {
                        int nilai = Integer.parseInt(scanner.nextLine().trim());
                        ht.delete(nilai);
                    } catch (NumberFormatException e) {
                        System.out.println("  [!] Input tidak valid. Masukkan angka bulat.");
                    }
                    break;

                case "3":
                    System.out.println("\n  -- CARI DATA --");
                    System.out.print("  Masukkan angka yang ingin dicari: ");
                    try {
                        int nilai = Integer.parseInt(scanner.nextLine().trim());
                        ht.search(nilai);
                    } catch (NumberFormatException e) {
                        System.out.println("  [!] Input tidak valid. Masukkan angka bulat.");
                    }
                    break;

                case "4":
                    ht.display();
                    break;

                case "5":
                    System.out.println("\n  Program selesai. Sampai jumpa!\n");
                    scanner.close();
                    return;

                default:
                    System.out.println("  [!] Pilihan tidak valid. Masukkan 1-5.");
            }
        }
    }
}
